File: 0036-valid-sudoku/0036-valid-sudoku.py
from collections import defaultdict

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = defaultdict(set)
        col = defaultdict(set)
        grid = defaultdict(set)

        ROWS = len(board)
        COLS = len(board[0])
        GRID_SIZE = 3

        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] != ".":
                    val = board[r][c]
                    g = (r // GRID_SIZE, c // GRID_SIZE)
                    
                    if val in row[r] or val in col[c] or val in grid[g]:
                        return False
                    
                    row[r].add(val)
                    col[c].add(val)
                    grid[g].add(val)

        return True

File: 0036-valid-sudoku/0036-valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    let rows = board.length, cols = board[0].length;
    let rowMap = {};
    let colMap = {};
    let subBoxMap = {};

    for( let r=0; r<rows; r++ ) {
        for( let c=0; c<cols; c++ ) {
            let val = board[r][c];

            if( val === "." )
                continue;
            
            if(!rowMap[r])
                rowMap[r] = new Set();
            
            if(!colMap[c])
                colMap[c] = new Set();

            let key = `${Math.floor(r/3)}-${Math.floor(c/3)}`;
            if(!subBoxMap[key])
                subBoxMap[key] = new Set();

            if(rowMap[r].has(val) || colMap[c].has(val) || subBoxMap[key].has(val))
                return false

            
            rowMap[r].add(val);
            colMap[c].add(val);
            subBoxMap[key].add(val);
        }
    }

    return true
};